src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -1,10 +1,393 @@
+            Ka = texture(map_Ka, UV).rgb;
+        else
+            Ka = PC.Ka;
+        if ((PC.extra&(map_Ks_MASK)) == (map_Ks_MASK))
+            Ks = texture(map_Ks, UV).rgb;
+        else
+            Ks = PC.Ks;
+        if ((PC.extra&(map_Ns_MASK)) == (map_Ns_MASK))
+            Ns = texture(map_Ns, UV).x;
+        else
+            Ns = PC.Ns;
+
+        color = Ka*Ia + Kd*NdotL;
+        switch (PC.extra&ILLUM_MODEL_MASK)
+        {
+        case 2:
+        case 3://2 + reflection map
+        case 4://3 with transparency (glass)
+        case 6:
+        case 8://reflection map
+        case 9://reflection map
+            color += Ks*pow(VdotR, Ns);
+            break;
+        case 5:
+        case 7:
+            color += Ks*pow(VdotR, Ns)*Fresnel_dielectric(PC.Ni, NdotL);
+            break;
+        default:
+            break;
+        }
+    }
+    else color = Kd;
+
+    OutColor = vec4(color, d);
+}
+#endif //_IRR_FRAG_MAIN_DEFINED_
+)";
+}
+
+
 using namespace irr;
 using namespace asset;
 
+static void insertShaderIntoCache(core::smart_refctd_ptr<ICPUSpecializedShader>& asset, const char* path, IAssetManager* _assetMgr)
+{
+    asset::SAssetBundle bundle({ asset });
+    _assetMgr->changeAssetKey(bundle, path);
+    _assetMgr->insertAssetIntoCache(bundle);
+}
+
+_IRR_STATIC_INLINE_CONSTEXPR const char* VERT_SHADER_NO_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/vertex_no_uv";
+_IRR_STATIC_INLINE_CONSTEXPR const char* VERT_SHADER_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/vertex_uv";
+_IRR_STATIC_INLINE_CONSTEXPR const char* FRAG_SHADER_NO_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/fragment_no_uv";
+_IRR_STATIC_INLINE_CONSTEXPR const char* FRAG_SHADER_UV_CACHE_KEY = "irr/builtin/shader/loaders/mtl/fragment_uv";
+
 CGraphicsPipelineLoaderMTL::CGraphicsPipelineLoaderMTL(IAssetManager* _am) : m_assetMgr{_am}
 {
+    //create vertex shaders and insert them into cache
+    {
+        std::string vs_nouv_source = VERT_SHADER_UV;
+        vs_nouv_source.insert(vs_nouv_source.find('\n') + 1ull,
+            R"(
+#define _IRR_VERT_MAIN_DEFINED_
+)"
+);
+        vs_nouv_source.insert(vs_nouv_source.size(), VERT_SHADER_NO_UV_MAIN);
+
+        auto vs_nouv_unspec = core::make_smart_refctd_ptr<ICPUShader>(vs_nouv_source.c_str());
+        auto vs_uv_unspec = core::make_smart_refctd_ptr<ICPUShader>(VERT_SHADER_UV);
+
+        ICPUSpecializedShader::SInfo specinfo({}, nullptr, "main", ICPUSpecializedShader::ESS_VERTEX);
+        auto vs_nouv = core::make_smart_refctd_ptr<ICPUSpecializedShader>(std::move(vs_nouv_unspec), ICPUSpecializedShader::SInfo(specinfo));
+        auto vs_uv = core::make_smart_refctd_ptr<ICPUSpecializedShader>(std::move(vs_uv_unspec), std::move(specinfo));
+
+        insertShaderIntoCache(vs_nouv, VERT_SHADER_NO_UV_CACHE_KEY, m_assetMgr);
+        insertShaderIntoCache(vs_uv, VERT_SHADER_UV_CACHE_KEY, m_assetMgr);
+    }
+    //create fragment shaders and insert them into cache
+    {
```

Comment:
what about PBR shaders?

Reply:
PBR shader exists for no reason and i can delete it just as well, its not used anywhere. All PBR params in MTL are non-standard (there's no `illum` in MTL that is using PBR params) and they are read and held in metadata just in case user wants to use them.